nidirect_workflow/src/Controller/AuditController.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace Drupal\nidirect_workflow\Controller;
+
+use Drupal\Core\DependencyInjection\ContainerInjectionInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\Controller\ControllerBase;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Class AuditController.
+ */
+class AuditController extends ControllerBase implements ContainerInjectionInterface {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * A logger instance.
+   *
+   * @var \Psr\Log\LoggerInterface
+   */
+  protected $logger;
+
+  /**
+   * The current user.
+   *
+   * @var \Drupal\Core\Session\AccountInterface
+   */
+  protected $account;
+
+  /**
+   * Creates a new ModerationStateConstraintValidator instance.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   * @param \Psr\Log\LoggerInterface $logger
+   *   The logger interface.
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The current user.
+   */
+  public function __construct(EntityTypeManagerInterface $entity_type_manager, LoggerInterface $logger, AccountInterface $account) {
+    $this->entityTypeManager = $entity_type_manager;
+    $this->logger = $logger;
+    $this->account = $account;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity_type.manager'),
+      $container->get('logger.factory')->get('nidirect_workflow'),
+      $container->get('current_user')
+    );
+  }
+
+  /**
+   * Content Audit.
+   *
+   * @return string
+   *   Return confirmation string.
+   */
+  public function contentAudit($nid) {
+    $msg = 'Invalid node id';
+    if (!empty($nid)) {
+      $node = $this->entityTypeManager()->getStorage('node')->load($nid);
+      if ($node) {
+        $audit_button_text = $this->config('nidirect_workflow.auditsettings')->get('audit_button_text');
+        $audit_confirmation_text = $this->config('nidirect_workflow.auditsettings')->get('audit_confirmation_text');
+        $msg = $this->t($audit_confirmation_text);
+        $msg .= "<div><a href='/nidirect_workflow/confirm_audit/$nid?destination=/node/$nid'>" . $this->t($audit_button_text) . "</a></div>";
```

Comment:
HTML in the controller doesn't feel right to me. I'd consider using a Link/Url building service here and move the divs etc out into a theme function/template.

Reply:
Hmm, well, I kind of see what you mean but I'm not sure that I agree.

We are in a controller here and we are returning markup, so it is bound to be HTML isn't it?

If you have a look at DbUpdateController.php in core you'll see that there is plenty of HTML in there.

Reply:
Yeah, I'd use https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Link.php/class/Link/8.2.x
as it returns a render array that if it ever needs overridden it can easily be done.

Reply:
Ahh, yes, of course. Apologies Johan I misunderstood you, I thought that you were suggesting that _I_ should write a Link/Url building service! (Hence my overkill comment.)

I do agree with you both that I should use the Drupal Link object, I'll go ahead and do that.